### ISSUE TITLE
delete application did not delete CRDs, nor update them and did not inform about this

### ISSUE THREAD

**Issue by KlavsKlavsen** [argoproj/argo-cd#6078]
Running ArgoCD 2.0 - when I installed my chart (velero 2.15.0) it installed some CRDs as well.
They show up fine in argocd list of resource for velero application.
I then deleted application argocd - but even though the UI shows application is gone - it did not actually delete the CRD objects :(
I then updated chart to 2.18.2 - which has updated CRDs - but when sync'ing this with argocd (previous installation was deleted by argocd) - then it does not touch the CRDs.. it shows them just fine - but "kubectl get crd" - reveals the CRDs hasn't been changed since the 2.15.0 installation. (I ran 2.15.0 with 2 patched velero CRDs so I know those changed in 1.6.0 as my patch has not yet been merged upstream.)

If argocd cannot delete a CRD - it should IMHO clearly complain about this and NOT delete the application (as its resources is not fully deleted)
When a CRD is installed - and the CRD in helm chart does not match that - argocd should clearly show this. instead of saying ALL GREEN :(
a CRD not matching the desired state is a serious issue and argocd does nothing to show you this is the BAD state the application is on.

**Comment by alexmt** [comment]
This known not perfect behavior. Argo CD does not delete CRDs because it is like deleted database table and unwanted behavior in most cases. I agree there should be at least a warning.

So technically it is not a bug. Let's use this issue to enhance current behavior and warn users that CRDs wont be deleted.

**Comment by KlavsKlavsen** [comment]
And also it did not update CRDs.. thats even worse IMHO.. it means I can't trust any of the helm chart updates done (that has CRDs) - they just show up without a diff - even though there clearly was one :(
If it has just shown a diff and said to "delete crd manually and then sync" or whatever..

**Comment by rajivml** [comment]
we are also facing this exact same issue , where ever CRD's are involved (rook-ceph, prometheus stack for example) , there the app deletion is getting stuck because namespace deletion is getting stuck in terminating state because of crd's which aren't deleted

**Comment by KlavsKlavsen** [comment]
@alexmt IMHO it IS a bug that ArgoCD does NOT update CRDs either.. atleast when using Helm.. it means you actually CANNOT use Helm with ArgoCD for any application which has CRDs.. thats a pretty big limitation - and the UI doesn't even tell you about it.. it installs CRDs the first time - but doesn't update them as far as I can tell ?

**Comment by rouke-broersma** [comment]
@KlavsKlavsen are your crd's in the crd folder of the chart or in the template folder?

**Comment by KlavsKlavsen** [comment]
in the crd folder typicly..  Its not MY CRD's.. its the application I manage with argocd that has CRDs.. which is almost EVERY application.. its becoming very popular to use CRDs.. ArgoCds own CRDs has the same problem when ArgoCD is used to manage itself (which I do).

**Comment by KlavsKlavsen** [comment]
The Velero chart f.ex. https://github.com/vmware-tanzu/helm-charts/tree/main/charts/velero - but also look at argocd chart. Same issue.

**Comment by rouke-broersma** [comment]
Mhm I can confirm that argo does not perform CRD deletions, but our argo does update CRD's.

**Comment by KlavsKlavsen** [comment]
Also when Argocd manages it? it doesn't run helm. And why should argocd handle CRDs from argocd helm chart any differently than when its pointed to velero helm chart.? If you can point to what should be the cause - I'll gladly submit such a fix to velero and other CRD handling Helm charts..
And - the worst part here - is that I get ZERO alerts about the CRD in cluster isn't the SAME as the one generated from Helm.

**Comment by rouke-broersma** [comment]
Everything in our cluster is managed by argocd, nothing is done manually. Afaik argo does a `helm template --include-crds` which makes plain kubernetes manifests and does a `kubectl apply` and nothing else special.

**Comment by KlavsKlavsen** [comment]
You are indeed correct.. Just tested and it showed me the CRD diff just fine.. I'll try and reproduce with patched CRDs.. it seems to work fine for "normal upgrades" thankfully.

**Comment by morningspace** [comment]
Seeing the same issue when I was trying to deploy/undeploy rook-ceph chart, CRDs not deleted although the Argo Application has gone. Understood that this is expected based on the above discussion history. Can it be settings turned off by default, for example, in `argocd-cm` ConfigMap, so that it gives people an option after they clearly know the risks and need to be enabled explicitly

**Comment by iampranabroy** [comment]
@rouke-broersma Is there a way to achieve that if I want to delete certain CRD's as part of the deletion application from ArgoCD?

**Comment by rouke-broersma** [comment]
No, you have to delete them manually.

**Comment by jemag** [comment]
I feel like there should at least be a way to tell ArgoCD that you want to delete CRDs for specific applications. Having to do it manually for 100% of situations seems counter-productive

**Comment by iampranabroy** [comment]
Adding the `resources-finalizer.argocd.argoproj.io` helped in deleting the resources/CRD in my case. ArgoCD doc [here](https://argocd-applicationset.readthedocs.io/en/stable/Application-Deletion/)

``` 
metadata:
  finalizers:
    - resources-finalizer.argocd.argoproj.io
```

**Comment by rouke-broersma** [comment]
Are you absolutely certain argocd deleted your crd's? Because it should not, by design. Adding a finalizer should not change anything in the case of crds.

**Comment by durgell-bl** [comment]
I've added `resources-finalizer.argocd.argoproj.io` and it does not delete CRD's.

I've even used `.yaml` explicitly defined CRDs (not in a helm chart) and `argoCD` is definitely not able to delete the CRD's

**Comment by dllegru** [comment]
I've added resources-finalizer.argocd.argoproj.io and it does not delete CRD's.

I've even used .yaml explicitly defined CRDs (not in a helm chart) and argoCD is definitely not able to delete the CRD's

**Comment by dllegru** [comment]
> Adding the `resources-finalizer.argocd.argoproj.io` helped in deleting the resources/CRD in my case. ArgoCD doc [here](https://argocd-applicationset.readthedocs.io/en/stable/Application-Deletion/)
> 
> ```
> metadata:
>   finalizers:
>     - resources-finalizer.argocd.argoproj.io
> ```



I've added `resources-finalizer.argocd.argoproj.io` and it does not delete CRD's.

I've even used .yaml explicitly defined CRDs (not in a helm chart) and argoCD is definitely not able to delete the CRD's

**Comment by dllegru** [comment]
> This known not perfect behavior. Argo CD does not delete CRDs because it is like deleted database table and unwanted behavior in most cases. I agree there should be at least a warning.
> 
> So technically it is not a bug. Let's use this issue to enhance current behavior and warn users that CRDs wont be deleted.

IMO, this should not be the desired behaviour for resources managed by code, the aim of it is that reflects what is in the source of truth (the repo). If I want all my resources managed by code and full gitops, I should be able to delete them. 

The user should be able to decide him/herself if it's a wanted or unwanted behaviour (protecting the resources with "prune:true" or any other available protection mechanism). Argo should be agnostic on what kind of resources manages, why can we delete `namespaces` and not `CRDs`? I don't understand why it is making any difference in the kind of resources it manages. Users should have a full control on resources managed by code and not abstract out of them the decission on which kind of resources are "more important/dangerous" than others.

**Comment by alkdese** [comment]
I also have same issue with my own helm chart which was having only configmap. Once I deleted helm application, the application disappeared but config map still there 2 days after deletion. I have had SyncPolic.prune=true in the app. What am I doing wrong?

**Comment by rouke-broersma** [comment]
> I also have same issue with my own helm chart which was having only configmap. Once I deleted helm application, the application disappeared but config map still there 2 days after deletion. I have had SyncPolicy.prune=true in the app. What am I doing wrong? I use argo 2.3.4. Where to look? I have found that for some reason, when I have updated `argocd-server` deployment and redeployed it the old replicaset has not been deleted and pods were restarting quite often. Will repeat the test again and see if it works.

Did you set the finalizer on the application?

**Comment by alkdese** [comment]
> Did you set the finalizer on the application?

Nope. Do I need?

**Comment by rouke-broersma** [comment]
Without the finalizer, argocd will only remove resources that are removed from the application source. If the application is removed argocd will not remove resources.

**Comment by alkdese** [comment]
> Without the finalizer, argocd will only remove resources that are removed from the application source. If the application is removed argocd will not remove resources.

Many thanks for your answer. I still believe the default behaviour should be application deletion should clean up all its resources, unless I wanna keep them. I disagree with this https://argo-cd.readthedocs.io/en/stable/user-guide/app_deletion/#deletion-using-kubectl
I think it should be done opposite way, I mean I should specify finalizer not to do so() if I need, while in most cases I think it should delete by default. I do not see a reason why would I wanna delete app and keep resources. This is very unusual.
Is there a possibility to reconfigure ArgoCD to do opposite way?

**Comment by sidewinder12s** [comment]
If we want to treat CRDs as special, it should just be a flag in Argo similar to how helm has flags around how it deals with CRDs. There may be cases where you do want to keep them, but if I am remove an app that has CRDs and the controller for that CRD, I think it should do what I ask.

**Comment by yangtian9999** [comment]
2.5.3, still can not delete CRD automatically when delete Application CR.

**Comment by acuD1** [comment]
Hi ! Any tips to handle this case ? :)

**Comment by rouke-broersma** [comment]
Add

The

Finalizer

**Comment by sidewinder12s** [comment]
> Hi ! Any tips to handle this case ? :)

I've just had to go in manually and delete the CRDs using kubectl, they are at least fairly easy to catch since they have consistent naming and are not namespaced.

**Comment by traviisd** [comment]
If I want to delete an application, why can't I delete all components of that application? The point of a tool like this is to help manage deployments and such (I'm oversimplifying, yes), but this specific behavior is essentially trying to change the meaning of delete and is just plain confusing! We have to manually delete any of the CRDs even though there is an admissions webhook and controller handling what to do when the CRD is deleted. Let Kubernetes and downstream process handle what should happen, it's why K8s is so extensible; Argo is taking this extensibility away in this case.

**Comment by shinebayar-g** [comment]
Hmm after debugging weird issues of helm, server side apply and CRDs for hours, finally landed here.
I absolutely agree with above users.

1. Why there is no indication that ArgoCD doesn't delete CRDs? If there was at least some tooltip text or something somewhere, that would have been helpful. https://argo-cd.readthedocs.io/en/stable/user-guide/auto_sync/ documentation doesn't mention anything about it.
2. Why ArgoCD won't delete resources it created when Application is deleted? IMHO by default ArgoCD should delete every child resources and there could be some flag or checkbox that could be configured at Application level if that's not something you want. No need of manually adding finalizers. Current behavior feels backwards. This would be also consistent with `helm uninstall` user experience.

**Comment by SnoozeFreddo** [comment]
+1 on the workaround

**Comment by smirnov-mi** [comment]
I'm facing the same issue (CRDs not removed after deinstalling ArgoCD via Rancher 2.8.1 GUI)
Workaround - remove these manually:
 
kubectl api-resources -o name | grep argo
applications.argoproj.io
applicationsets.argoproj.io
appprojects.argoproj.io

**Comment by smirnov-mi** [comment]
> > Hi ! Any tips to handle this case ? :)
> 
> I've just had to go in manually and delete the CRDs using kubectl, they are at least fairly easy to catch since they have consistent naming and are not namespaced.

easy to locate, but they are namespaced (at least now, having installed ArgoCD as an app. via RancherGUI).

**Comment by sergeyshevch** [comment]
I guess some flag in the application can be good option here. Like If I as operator know that only this app will use CRD then I can enable it and cleanup CRDs

**Comment by pdeva** [comment]
bumping this up since its still an issue

**Comment by reggie-k** [comment]
**TLDR**:
Argo CD does not track the CRDs it deploys (does not add resource tracking label / annotation), **by design**, to prevent their deletion from the cluster (either explicit deletion from Git or implicit deletion by deleting the Application) - this is done since a CRD is a resource, the deletion of which can have broad scope implications. This behavior does not align well with GitOps concept when a user deletes the CRD from Git or deletes the Application (if it has a finalizer).
However, deleting the CRD explicitly from Argo CD UI **does** delete it from the cluster, but the UI is does not reflect this immediately, only upon the next refresh it is shown as a diff.

Changes in the CRD in Git, however, are reflected as a diff, aligning well with GitOps concepts.

Upon deleting a CRD from Git, it is correctly deleted from Argo CD UI, but in fact the CRD remains in the cluster, which can be misleading. Also, the experience of deleting a CRD in multiple ways does not have a stream-line behavior (possible to delete from the UI but impossible to delete when removing from Git and impossible to delete when removing the managing App).

**Short-term action items:**
- Document the existing behavior better

**Possible longer-term solutions:**
- New feature to enable deletion (change in behavior): add an Application level / global config sync option for deleting CRDs and make the CRDs tracked and deleted like any other resource when the option is enabled.
- New feature to warn about non-deletion (no change in behavior): If the Application has a CRD, warn upon deleting the Application /  pruning the CRD (that was removed from Git) that the CRD will not actually be removed from the cluster
- Change in existing explicit deletion behavior - make the explicit resource deletion (UI and CLI) behave the same way as pruning / deleting an Application, regardless of which previous direction was chosen
